File: algorithms.js
function insertionSort(array, descendingOrder = false) {
    for (let i = 1; i < array.length; i++) {
        let key = array[i];
        let j = i - 1;
        if (descendingOrder) {
            while (j >= 0 && array[j] < key) {
                array[j + 1] = array[j];
                j = j - 1
            }
            array[j + 1] = key
        } else {
            while (j >= 0 && array[j] > key) {
                array[j + 1] = array[j];
                j = j - 1
            }
            array[j + 1] = key
        }
    }

    return array;
}

function selectionSort(array, descendingOrder = false) {
    for (let i = 0; i < array.length - 1; i++) {
        let m = i;
        let j = i + 1;
        if (descendingOrder) {
            while (j < array.length) {
                if (array[j] > array[m]) {
                    m = j
                }
                j = j + 1
            }
        } else {
            while (j < array.length) {
                if (array[j] < array[m]) {
                    m = j
                }
                j = j + 1
            }
        }
        let buffer = array[i];
        array[i] = array[m];
        array[m] = buffer;
    }
    return array;
}

function mergeSort(array, descendingOrder = false) {
    if (array.length > 1) {
        let mid = array.length / 2;
        let left_array = array.slice(0, mid);
        let right_array = array.slice(mid);

        mergeSort(left_array);
        mergeSort(right_array);

        let i = 0;
        let j = 0;
        let k = 0;

        while (i < left_array.length && j < right_array.length) {
            if (left_array[i] < right_array[j]) {
                array[k] = left_array[i];
                i += 1;
            } else {
                array[k] = right_array[j];
                j += 1;
            }
            k += 1;
        }

        while (i < left_array.length) {
            array[k] = left_array[i];
            i += 1;
            k += 1;
        }

        while (j < right_array.length) {
            array[k] = right_array[j];
            j += 1;
            k += 1;
        }
    }
    if (descendingOrder) {
        return array.reverse();
    } else {
        return array
    }
}

function bubbleSort(array, descendingOrder = false) {
    for (let i = 0; i < array.length - 1; i++) {
        for (let j = 0; j < array.length - i - 1; j++) {
            if (descendingOrder) {
                if (array[j] < array[j + 1]) {
                    let buffer = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = buffer;
                }
            } else {
                if (array[j] > array[j + 1]) {
                    let buffer = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = buffer;
                }
            }
        }
    }
    return array;
}

function findMaxSubarray(array) {
    let best_sum = 0;
    let current_sum = 0;

    for (let i = 0; i < array.length; i += 1) {
        current_sum = Math.max(0, current_sum + array[i]);
        best_sum = Math.max(best_sum, current_sum);
    }

    return best_sum;
}

function test() {
    console.log("Testing insertion sort");
    console.log("______________________");
    console.log(insertionSort([5, 3, 4, 2, 1, 2, 3, 5, 2], true));
    console.log(insertionSort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]));
    console.log(insertionSort(["a", "b", "d", "a", "h", "a"]));
    console.log(insertionSort(["a", "b", "d", "a", "h", "a"], true));
    console.log("______________________");

    console.log("Testing selection sort");
    console.log("______________________");
    console.log(selectionSort([5, 3, 4, 2, 1, 2, 3, 5, 2], true));
    console.log(selectionSort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]));
    console.log(selectionSort(["a", "b", "d", "a", "h", "a"]));
    console.log(selectionSort(["a", "b", "d", "a", "h", "a"]), true);
    console.log("______________________");

    console.log("Testing merge sort");
    console.log("______________________");
    console.log(mergeSort([5, 3, 4, 2, 1, 2, 3, 5, 2], true));
    console.log(mergeSort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]));
    console.log(mergeSort(["a", "b", "d", "a", "h", "a"]));
    console.log(mergeSort(["a", "b", "d", "a", "h", "a"], true));
    console.log("______________________");

    console.log("Testing bubble sort");
    console.log("______________________");
    console.log(bubbleSort([5, 3, 4, 2, 1, 2, 3, 5, 2], true));
    console.log(bubbleSort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]));
    console.log(bubbleSort(["a", "b", "d", "a", "h", "a"]));
    console.log(bubbleSort(["a", "b", "d", "a", "h", "a"], true));
    console.log("______________________");

    console.log("Testing find maximum subarray");
    console.log("______________________");
    console.log(findMaxSubarray([5, 3, -4, 2, 1, -2, 3, 5, 2]));
    console.log(findMaxSubarray([1, 2, -4, 2, 2, -1, 5, 6, 6, 9]));
    console.log("______________________");
}

test();

File: algorithms.py
def insertion_sort(array, descending_order=False):
    for i in range(1, len(array)):
        key = array[i]
        j = i - 1
        if descending_order:
            while j >= 0 and array[j] < key:
                array[j + 1] = array[j]
                j = j - 1
            array[j + 1] = key
        else:
            while j >= 0 and array[j] > key:
                array[j + 1] = array[j]
                j = j - 1
            array[j + 1] = key
    return array


def selection_sort(array, descending_order=False):
    for i in range(len(array) - 1):
        m = i
        j = i + 1
        if descending_order:
            while j < len(array):
                if array[j] > array[m]:
                    m = j
                j = j + 1
        else:
            while j < len(array):
                if array[j] < array[m]:
                    m = j
                j = j + 1
        array[i], array[m] = array[m], array[i]
    return array


def merge_sort(array, descending_order=False):
    if len(array) > 1:
        mid = len(array) // 2
        left_array = array[:mid]
        right_array = array[mid:]

        merge_sort(left_array)
        merge_sort(right_array)

        i = j = k = 0

        while i < len(left_array) and j < len(right_array):
            if left_array[i] < right_array[j]:
                array[k] = left_array[i]
                i += 1
            else:
                array[k] = right_array[j]
                j += 1
            k += 1

        while i < len(left_array):
            array[k] = left_array[i]
            i += 1
            k += 1

        while j < len(right_array):
            array[k] = right_array[j]
            j += 1
            k += 1
    if descending_order:
        return list(reversed(array))
    else:
        return array


def bubble_sort(array, descending_order=False):
    for i in range(len(array) - 1):
        for j in range(len(array) - i - 1):
            if descending_order:
                if array[j] < array[j + 1]:
                    array[j], array[j + 1] = array[j + 1], array[j]
            else:
                if array[j] > array[j + 1]:
                    array[j], array[j + 1] = array[j + 1], array[j]
    return array


def find_max_subarray(array):
    best_sum = 0
    best_start = best_end = 0
    current_sum = 0

    for current_end, x in enumerate(array):
        if current_sum <= 0:
            current_start = current_end
            current_sum = x
        else:
            current_sum += x

        if current_sum > best_sum:
            best_sum = current_sum
            best_start = current_start
            best_end = current_end + 1

    return best_sum, best_start, best_end


def test():
    print("Testing insertion sort")
    print("______________________")
    print(insertion_sort([5, 3, 4, 2, 1, 2, 3, 5, 2], True))
    print(insertion_sort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]))
    print(insertion_sort(["a", "b", "d", "a", "h", "a"]))
    print(insertion_sort(["a", "b", "d", "a", "h", "a"], True))
    print("______________________")

    print("Testing selection sort")
    print("______________________")
    print(selection_sort([5, 3, 4, 2, 1, 2, 3, 5, 2], True))
    print(selection_sort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]))
    print(selection_sort(["a", "b", "d", "a", "h", "a"]))
    print(selection_sort(["a", "b", "d", "a", "h", "a"], True))
    print("______________________")

    print("Testing merge sort")
    print("______________________")
    print(merge_sort([5, 3, 4, 2, 1, 2, 3, 5, 2], True))
    print(merge_sort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]))
    print(merge_sort(["a", "b", "d", "a", "h", "a"]))
    print(merge_sort(["a", "b", "d", "a", "h", "a"], True))
    print("______________________")

    print("Testing bubble sort")
    print("______________________")
    print(bubble_sort([5, 3, 4, 2, 1, 2, 3, 5, 2], True))
    print(bubble_sort([1, 2, 4, 2, 2, 1, 5, 6, 6, 9]))
    print(bubble_sort(["a", "b", "d", "a", "h", "a"]))
    print(bubble_sort(["a", "b", "d", "a", "h", "a"], True))
    print("______________________")

    print("Testing find maximum subarray")
    print("______________________")
    print(find_max_subarray([5, 3, -4, 2, 1, -2, 3, 5, 2]))
    print(find_max_subarray([1, 2, -4, 2, 2, -1, 5, 6, 6, 9]))
    print("______________________")


test()
